File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        int n=s.size();
        vector<int>hash(3,-1);
        int count=0;
        for(int i=0;i<n;i++){
            hash[s[i]-'a']=i;
            count=count+(1+min(hash[0],min(hash[1],hash[2])));
        }

        return count;
    }
};

File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        res = 0
        n = len(s)

        i = 0
        j = 0
        mp = [0]*3

        for j in range(n):
            mp[ord(s[j])-ord('a')]+=1

            while(mp[0]>0 and mp[1]>0 and mp[2]>0):
                res += n-j
                mp[ord(s[i])-ord('a')]-=1
                i+=1
            
        
        return res